scip-java/src/main/scala/com/sourcegraph/scip_java/buildtools/BazelBuildTool.scala

```diff
@@ -13,15 +26,216 @@ class BazelBuildTool(index: IndexCommand) extends BuildTool("Bazel", index) {
+              .call(check = false, stdout = os.Inherit, stderr = os.Inherit)
+          }
+          index
+            .app
+            .error(
+              s"""To reproduce the failed Bazel command without scip-java, run the following command:
+                 |
+                 |  bazel build ${targetSpecs.mkString(" ")}
+                 |
+                 |To narrow the set of targets to index or pass additional flags to Bazel, include extra arguments index after -- like below:
+                 |
+                 |  scip-java index --bazel-scip-java-binary=... -- //custom/target --sandbox_debug
+                 |""".stripMargin
+            )
+          return commandResult.exitCode
+        }
+
+        // Final step after running the aspect: aggregate all the generated `*.scip` files into a single index.scip file.
+        // We have to do this step outside of Bazel because Bazel does not allow actions to generate outputs outside
+        // of the bazel-out directory. Ideally, we should be able to implement the aggregation step inside Bazel
+        // and only copy the resulting index.scip file into the root of the workspace. However, to begin with, we
+        // walk the bazel-bin/ directory to concatenate the `*.scip` files even if this is not the idiomatic way to
+        // do it (in general, scanning the bazel-bin/ directory is a big no no).
+        Files.deleteIfExists(index.finalOutput)
+        Files.createDirectories(index.finalOutput.getParent)
+        val scipPattern = FileSystems.getDefault.getPathMatcher("glob:**.scip")
+        val bazelOut = index.workingDirectory.resolve("bazel-out")
+        if (!Files.exists(bazelOut)) {
+          index
+            .app
+            .error(
+              s"doing nothing, the directory $bazelOut does not exist. " +
+                s"The most likely cause for this problem is that there are no Java targets in this Bazel workspace. " +
+                s"Please report an issue to the scip-java issue tracker if the command `bazel query 'kind(java_*, //...)'` returns non-empty output."
+            )
+          return 0
+        }
+        val bazelOutLink = Files.readSymbolicLink(bazelOut)
+        Files.walkFileTree(
+          bazelOutLink,
+          new SimpleFileVisitor[Path] {
+            override def visitFile(
+                file: Path,
+                attrs: BasicFileAttributes
+            ): FileVisitResult = {
+              if (scipPattern.matches(file)) {
+                val bytes = Files.readAllBytes(file)
+                Files.write(
+                  index.finalOutput,
+                  bytes,
+                  StandardOpenOption.APPEND,
+                  StandardOpenOption.CREATE
+                )
+              }
+              super.visitFile(file, attrs)
+            }
+          }
+        )
+        0
+    }
+  }
+
+  /**
+   * Callback to `os.ProcessOutput.Readlines` that processes the stderr output
```

Comment:
For this and the various tree walking operations - is there a reason we can't just bring in `os-lib`? Seems like the cli has plenty of dependencies as it is, and it'd be a lot more readable in cases like this to just use `os.walk`

Reply:
Unless I'm misreadingh what you mean by `os.ProcessOutput.Readlines` - this? https://github.com/com-lihaoyi/os-lib/blob/d0661198c915ca037591067ce8967275baca6aae/os/src-jvm/SubProcess.scala#L218

Reply:
Wait, Moped already used os-lib :D 

I'm confused about the use of file visitors then

Reply:
For the record, the equivalent code (I believe) is

```scala
        val bazelOutLink = Files.readSymbolicLink(bazelOut)
        val out = os.Path(index.finalOutput)

        os.walk(
            os.Path(bazelOutLink),
            skip = path => scipPattern.matches(path.toNIO)
          )
          .foreach { p =>
            os.write.append(out, os.read.stream(p))
          }
```

Reply:
Old habits die hard. I'll admit that looks very nice, esp. knowing how it uses generators under the hood

Reply:
I tried running it and hit on
```
java.io.IOException: Is a directory
	at java.base/sun.nio.ch.FileDispatcherImpl.read0(Native Method)
```
because there are many directories with the `.scip` extension, and `os.walk` visits both files and directories. It's possible to `skip = path => os.isFile(path) && ...` but that requires an extra fsstat on every visited file